### DISCUSSION TITLE
JSON and Wish List

### DISCUSSION THREAD

**Discussion by giovannicoppola** [mkb79/audible-cli#57] [Q&A] [answered]
Very nice package! 💯 Is there a way to retrieve all the library details in JSON format? I can convert the .tsv file to JSON, but I was wondering if it is possible to do that directly.

Second, is there a way to download the Audible Wish List?
Thanks.

**Comment by mkb79** [comment]
> Very nice package! 💯 Is there a way to retrieve all the library details in JSON format? I can convert the .tsv file to JSON, but I was wondering if it is possible to do that directly.

The Audible server response is JSON. So it is easy to output this data directly!

> Second, is there a way to download the Audible Wish List? Thanks.

I can request the wishlist data from the Audible API. However, in order to download a file, it must first be in your own library.

**Reply by giovannicoppola** [reply]
that would be great! I would like to create an Alfred workflow to quickly query my library and wishlist, something that is typically cumbersome on the website

**Comment by mkb79** [comment]
@giovannicoppola
The development branch can now export your library as tsv, csv or json. In the next step I will implement the wishlist.

**Reply by giovannicoppola** [reply]
@mkb79 I was wondering if the wishlist is still on your roadmap?

**Comment by mkb79** [comment, marked as answer]
@giovannicoppola I‘ve implement the wishlist in the current development branch. You can use it when you clone it. Or you must wait until I release v0.1.0 to Pypi.

**Reply by giovannicoppola** [reply]
awesome! I will give it a try

**Comment by mkb79** [comment]
[Here](https://github.com/mkb79/audible-cli/blob/development/CHANGELOG.md) you can find the changelog for the development branch!

**Reply by giovannicoppola** [reply]
I don't know enough to use the devel branch ☹️, I will wait for the pip release. thanks!

**Reply by mkb79** [reply]
Simply use `git clone --branch development https://github.com/mkb79/audible-cli`! Then switch into the cloned dir and run `pip install --upgrade .`!

**Comment by giovannicoppola** [comment]
awesome!!! I got the wishlist. 
<img width="411" alt="image" src="https://user-images.githubusercontent.com/5774411/156898463-21ae853e-f9a7-4727-aac1-d04d673c573c.png">

~~How do I get the json output? can one have additional fields, or is it the exact same as tsv?~~ figured it out 🤓

**Reply by giovannicoppola** [reply]
@mkb79 workflow drafted, check out a preview [here](https://github.com/giovannicoppola/alfred-audible)

**Reply by mkb79** [reply]
@giovannicoppola I've got a 404 when I press the link!

**Reply by giovannicoppola** [reply]
oops, it is public now

**Reply by mkb79** [reply]
Looks very nice. If you need some more features (e.g. adding/removing an item to the wishlist) please tell me!

**Reply by giovannicoppola** [reply]
will do! Note that I cache table and images at user-defined intervals, so I don't hit audible all the time. Adding to the wishlist would be awesome, but don't you need to query the whole audible database to do that? Is that even possible?

**Reply by mkb79** [reply]
Using the API you can add/remove an item to the wishlist if you have the asin. You can search the catalog providing some keywords. This way you can get the asin for the item you want to add!

**Reply by giovannicoppola** [reply]
yes I can see a use case in which the user finds a book online and wants to 1) check if it is on audible and 2) add to the wishlist, via Alfred. will the search too be possible through the API?

**Reply by mkb79** [reply]
Yes, of course. You can use to fetch the catalog providing some keywords. You can use the API tosearch for an author or a narrator too. [Here](https://audible.readthedocs.io/en/latest/misc/external_api.html#get-1-0-catalog-products) you can find some information about the catalog/products endpoint!

**Reply by giovannicoppola** [reply]
thanks! I might need some hand holding to use the python module directly. right now I just launch the CLI version

**Reply by mkb79** [reply]
I can add a command to `audible-cli` which let you use the API directly such as wget goes! So you don’t have to do with authentication/different marketplaces and so on. Instead of a url you provide the method (GET,POST,…), the endpoint, the params of the request and a body on POST requests and it gives you the the JSON response!

**Reply by giovannicoppola** [reply]
or, if you have a couple of examples searching the database using the package and `urllib` I think I can adapt and implement.. [this is](https://github.com/giovannicoppola/alfred-ticker/blob/main/source/ticker.py) my (low) level of expertise

**Reply by mkb79** [reply]
If you want to use urllib you have to implement the authentication and authorization part yourself. This can be very tricky.

**Reply by giovannicoppola** [reply]
no I was thinking to leave everything as is now (authentication via CLI), but just search and add to wishlist using `urllib`. It should work if you are already authenticated, right?

**Comment by mkb79** [comment]
Ah, you mean using urllib to make a request to `audible-cli` which make the request to the API and gives you the response? I can running a Flask Server (with Rest Implantation) via a subcommand. Then you can make requests to the Flask server! But this need more time!

**Reply by giovannicoppola** [reply]
that's fine, I was just trying to save you work :) if you think the wget implementation you recommended is the fastest way, I will wait for that. thanks for your patience!

**Reply by mkb79** [reply]
I think I will implement the API command (like wget) as a built-in command. This needs no additional packages. 

Then I will write a plugin to run a Flask Rest server. This reduce depending for users, which don’t need this! Then you can run multiple Flask server on different ports for different profiles!

**Reply by mkb79** [reply]
But this will come after my work on v0.1.0 is finished. I need to implement one thing and then I will release it.

**Reply by giovannicoppola** [reply]
quick question: I am trying to go through the installation steps and write the documentation from a fresh computer as I have tried many things on my main computer. After `pip3 install audible-cli`, the package is installed, but I can't launch `audible-quickstart` from any folder, as I do on my main computer (I guess it is not added to the path). On my main computer I was able to launch audible from anywhere even before installing the development branch. Any guesses on what I am missing?

**Reply by mkb79** [reply]
You are using venv? Can you try `audible quickstart` instead?

**Reply by giovannicoppola** [reply]
yes, I tried both 
`audible quickstart`
`zsh: command not found: audible`

**Reply by mkb79** [reply]
I does not know if the folder is in your path. But you can try `python3 -m audible_cli quickstart`. This should work too

**Reply by giovannicoppola** [reply]
that works! for some reason on my other computer `audible library list` works from everywhere while on this 'fresh' computer I need to enter `python3 -m audible_cli library list`. Sorry for the noob questions!

**Reply by mkb79** [reply]
Which OS you are using. You can search for a file named `audible` or `audible.exe` (depending on your OS).

**Reply by giovannicoppola** [reply]
macOS big Sur (the main computer where everything works is Monterey). I can launch from the terminal but not from Alfred (I get these errors below), I will try to figure it out. It could be a conflict between Python3 installations. Typically it is solved by adding 
`export PATH=/opt/homebrew/bin:/usr/local/bin:$PATH` but no luck so far
```
[14:27:48.076] ERROR: alfred-audible[[Script Filter](alfredpreferences:workflows%3Eworkflow%3Egiovanni-audible%3E0AC8F1DB-D086-42D0-8052-72A969F42ACD)] Code 1: Traceback (most recent call last):
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py", line 185, in _run_module_as_main
    mod_name, mod_spec, code = _get_module_details(mod_name, _Error)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py", line 144, in _get_module_details
    return _get_module_details(pkg_main_name, error)
  File "/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.8/lib/python3.8/runpy.py", line 111, in _get_module_details
    __import__(pkg_name)
  File "/Users/giovanni.coppola/Library/Python/3.8/lib/python/site-packages/audible_cli/__init__.py", line 5, in <module>
    from .cli import main, quickstart
  File "/Users/giovanni.coppola/Library/Python/3.8/lib/python/site-packages/audible_cli/cli.py", line 6, in <module>
    from .cmds import build_in_cmds, cmd_quickstart
  File "/Users/giovanni.coppola/Library/Python/3.8/lib/python/site-packages/audible_cli/cmds/__init__.py", line 3, in <module>
    from . import (
  File "/Users/giovanni.coppola/Library/Python/3.8/lib/python/site-packages/audible_cli/cmds/cmd_activation_bytes.py", line 2, in <module>
    from audible.activation_bytes import (
ModuleNotFoundError: No module named 'audible.activation_bytes'; 'audible' is not a package
```

**Reply by mkb79** [reply]
Maybe you can use Pythons venv module to solve this issue?

**Reply by giovannicoppola** [reply]
update: in the working computer (number 1) audible is in /usr/local/bin; in the second computer it is in `/System/Volumes/Data/Users/giovanni.coppola/Library/Python/3.8/bin/` and it works if I launch from there (both from terminal and Alfred). I tried a third computer fairly new (OS Monterey) and it fails while installing `Pillow`

**Comment by mkb79** [comment]
@giovannicoppola 
I've released the [api_release](https://github.com/mkb79/audible-cli/tree/api_command) branch. You can test this out. It contains a proof-of-concept without any exception handling. Please tell me, if this is something you can work with. The Flask-Plugin will come in some days.

To use it this way:
- `audible api` is the command name (maybe I change this?)
- each query parameter must be provided with the `-p/--param` option, like `- p response_groups=product_desc,media,sample`. Yo can chain them with multiple `-p` options like `-p response_groups=media -p num_results=5`.
- for POST or PUT requests you can provide the body (as json) with the `-b` option
- at least you have to provide the endpoint like `library`, `wishlist` or `catalog/products`

Feel free to report me issues, if you find some, or feature requests.

**Reply by giovannicoppola** [reply]
this is beautiful @mkb79. I am able to browse the catalog and retrieve the options you listed. a couple of questions:
1. I see that the total number of records is 65,422. I see there is a 50 download limit, but any chance the whole database (or at least title, author, narrator, asin etc) can be downloaded every month or so for fast querying? After all this can only be result in more business for Audible... 
2.  how do I search? if you have an example I can go from there. 
3. unrelated: have you ever looked into doing something similar for [libby](https://www.overdrive.com/apps/libby)? That would also benefit from a better interface/search...

**Comment by mkb79** [comment]
@giovannicoppola 

1. You can use the `page` param to retrieve more items. But one user reports to me, that after some pages (I have 40 in my mind?!) the fetched items are always the same. So there will be a limit for doing this.
2. To search you can use the `keyword=…` parameter. Or if you know what you searching for you can use the `title`, `author` or `narrator` paramaters. Maybe you can mix them too?!
3. I have many ideas, what I would do with this project. A better interface are one of them. But I'm working alone on this project. So I can’t realize all my ideas. But help is welcome every time!

**Reply by mkb79** [reply]
I‘ve added the `api` command to the development branch. So this will be released with v0.1.0!

**Reply by giovannicoppola** [reply]
they all work except keyword? 
```
audible api catalog/products -p num_results=1 -p response_groups=product_desc                     
{'product_filters': [], 'products': [{'asin': '1524779261', 'subtitle': 'An Easy & Proven Way to Build Good Habits & Break Bad Ones', 'title': 'Atomic Habits'}], 'response_groups': ['product_desc', 'always-returned'], 'total_results': 64954}

audible api catalog/products -p num_results=1 -p response_groups=product_desc -p keyword=churchill 
{'product_filters': [], 'products': [{'asin': '1524779261', 'subtitle': 'An Easy & Proven Way to Build Good Habits & Break Bad Ones', 'title': 'Atomic Habits'}], 'response_groups': ['product_desc', 'always-returned'], 'total_results': 64954}

```

**Reply by mkb79** [reply]
@giovannicoppola You have to use `keywords` instead of `keyword` ;)

**Reply by giovannicoppola** [reply]
duh, thanks. and is there an endpoint that returns also author, narrator, and the little summary?

**Reply by mkb79** [reply]
You have to provide the correct response_group to get this info. 

Known `response_groups` values for the catalog/product endpoint are: `contributors, customer_rights, media, price, product_attrs, product_desc, product_extended_attrs, product_plan_details, product_plans, rating, sample, sku, series, reviews, ws4v, relationships, review_attrs, categories, category_ladders, claim_code_url, in_wishlist, listening_status, periodicals, provided_review, product_details`! You can select multiple response_groups and separate them with a comma!

**Comment by mkb79** [comment]
Here is my first attempt for a FlaskServer

```python
import logging

import click
from audible import Client
from flask import Flask
from flask.cli import FlaskGroup, ScriptInfo
from flask_restful import Api, Resource, request


logger = logging.getLogger("audible_cli.cmds.cmd_flask")

audible_cli_session = None
audible_cli_client = None


class RESTapp(Resource):

    @staticmethod
    def get(path=""):
        args = request.args
        r = audible_cli_client._request("GET", path, params=args)
        return r

    @staticmethod
    def post(path=""):
        json_body = request.get_json()
        r = audible_cli_client._request("POST", path, json=json_body)
        return r

    @staticmethod
    def put(path=""):
        json_body = request.get_json()
        r = audible_cli_client._request("PUT", path, json=json_body)
        return r

    @staticmethod
    def delete(path=""):
        r = audible_cli_client._request("DELETE", path)
        return r


def create_app():
    app = Flask(__name__)
    api = Api(app)
    api.add_resource(RESTapp, '/', '/<path:path>')
    return app


@click.group(
    "flask-server",
    cls=FlaskGroup,
    create_app=create_app,
    load_dotenv=False
)
@click.pass_context
def cli(ctx):
    """Using Flask to make request to the API"""

    # There can only be one `obj` in ctx. FlaskGroup needs his ScriptInfo
    # The audible_cli.session obj must be keep for further purposes
    global audible_cli_session
    global audible_cli_client
    audible_cli_session = ctx.obj
    audible_cli_client = Client(auth=audible_cli_session.auth)
    obj = ScriptInfo(
        create_app=create_app,
        set_debug_flag=ctx.command.set_debug_flag
    )
    ctx.obj = obj
```

Copy this to file named `cmd_flask-server.py` in the audible-cli plugin dir. Then install `flask-restful` in your venv or something else with `pip install flask-restful`. You can start it with `audible flask-server run --host 0.0.0.0`. It support the full Flask run command with his parameters!

**Reply by mkb79** [reply]
Or you can clone and install it from https://github.com/mkb79/audible-cli-flask

**Reply by giovannicoppola** [reply]
hi @mkb79 thanks for this. I am now able (with `audible api`) to query the catalog and return audiobook details, which is neat!! I haven't tried but I understand I should be able to add them to my wishlist as well. 

Is the flask server needed to alleviate the inconsistencies in installation I am observing across computers?

**Reply by mkb79** [reply]
The flask-server command works like the api command. The advantage of the flask-server command is, you can run them on one computer (with Python installed) and use a second computer to request the AudibleAPI using the Flask server. You can use a browser or any other APP, which makes http requests, for this. This reduce dependencies for the second computer.

**Reply by giovannicoppola** [reply]
catalog search [demo](https://github.com/giovannicoppola/alfred-audible). you can't set sorting criteria, correct?

**Reply by mkb79** [reply]
You can try the `sort_by` param like sort_by: [-Author, -DateAdded, -Price, -Rating, -Title, Author, DateAdded, Price, Rating, Title]

**Reply by mkb79** [reply]
Audible API is non public. So there is no official documentation about available endpoints, params and message bodys! The Audible app for Windows, Android or iOS uses them. So all information are obtained via HTTP sniffing!

**Comment by giovannicoppola** [comment]
got it thanks!
did you mean `-p sort_by:Author`? That raises an `Uncaught Exception`

**Comment by mkb79** [comment]
You have to use a `=` instead of `:` as separator. The catalog/products endpoint don’t have a `sort_by` param if found out now. But you can use products_sort_by: [-ReleaseDate, ContentLevel, -Title, AmazonEnglish, AvgRating, BestSellers, -RuntimeLength, ReleaseDate, ProductSiteLaunchDate, -ContentLevel, Title, Relevance, RuntimeLength]

**Comment by giovannicoppola** [comment]
Hi @mkb79 I am preparing the documentation and instructions to install `alfred-audible`. In general, alfred workflow developers try to streamline the process as much as possible, reduce the burden on the end-user, take care of dependancies etc. I can't figure out what I did on my desktop that added audible-cli to the path, so I will point my script to `~/Library/Python/3.8/bin/audible`. Is this the right way to do it or is there some system shortcut that points automatically to the most recent Python installation? Will users need to update as they change Python version? Thank you for your patience and guidance! As you can tell I am a beginner. 🤓

**Reply by giovannicoppola** [reply]
hello @mkb79 I see you released version 0.1.0!! I will proceed as I outlined in the post above, unless you have a different suggestion. This topic was recently discussed in a [thread on the Alfred forum](https://www.alfredforum.com/topic/18087-currency-converter/?do=findComment&comment=93539) and the recommendation is to bundle everything, however I don't know how to launch `audible-cli` at the command line without installing it. If you agree I will just ask users to install and then I will access using path/to/python

**Reply by mkb79** [reply]
> hello @mkb79 I see you released version 0.1.0!! I will proceed as I outlined in the post above, unless you have a different suggestion. This topic was recently discussed in a [thread on the Alfred forum](https://www.alfredforum.com/topic/18087-currency-converter/?do=findComment&comment=93539) and the recommendation is to bundle everything, however I don't know how to launch `audible-cli` at the command line without installing it. If you agree I will just ask users to install and then I will access using path/to/python

Maybe you can try the [standalone](https://github.com/mkb79/audible-cli/releases/download/latest/audible_mac.zip) binary for audible-cli. This should bring you all you need in one file (if you do not use your own plugins)!

**Comment by giovannicoppola** [comment]
ooh, do you now have a macos standalone? The link seems dead

**Reply by mkb79** [reply]
Yeah. I‘m using GithubActions and pyinstaller to create this. The link is working for me. Then visit https://github.com/mkb79/audible-cli/releases and download the binary for macOS.

**Reply by giovannicoppola** [reply]
hmm is it supposed to work like the package after installation? I downloaded, unzipped, and launched but no luck. 
```
giovanni@Giovannis-iMac Downloads % audible
zsh: command not found: audible
```
with `open audible` I get a window with the usage info, then exit

**Reply by mkb79** [reply]
It should run like before. Please try `open audible --version`. It should give you the version info for `audible-cli`. How long does it takes to start the app on macOS?

**Reply by mkb79** [reply]
Instead using the `open` command please try `./audible --version` from the directory where the binary is.

**Reply by giovannicoppola** [reply]
`./audible` works normally, thanks. Version is 0.1.0. Not extensive testing, but it seems considerably slower

**Reply by mkb79** [reply]
The binary have to be extracted every time, before it can be executed. On linux this takes only some milliseconds extra time. On Windows machines it depends on the Anti-Virus-Scanner and some other things. Maybe it‘s the same problem on macOS!

As a solution for Windows, I've added a `onedir` in addition to the `onefile` binary. This reduce startup time extremely. I will provide such a `onedir` solution for macOS in the next release, too.

**Reply by giovannicoppola** [reply]
ok, happy to help you troubleshoot. this is the output of `time`:
`./audible wishlist list  0.57s user 0.15s system 10% cpu 7.216 total`

**Reply by mkb79** [reply]
Thank you for this information. I‘ve added the `onedir` binary creation for macOS! The build process will first done automatically on my next release. Then you can test the speed again. 

In v0.1.0 you find the new `api` command. With these you can make custom requests to the AudibleAPI and it will give you the plain JSON output. So you can build your own API requests.

**Comment by giovannicoppola** [comment]
Hi @mkb79 , I hope you are well! I am back at this project and would like to release a first version of the workflow. I would like to bundle your package with my workflow (in the `lib` directory) with this command (requirements.txt = audible-cli:

`pip3 install -r requirements.txt --target=lib`

but I get the following error: 
`ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
httplib2 0.19.1 requires pyparsing<3,>=2.4.2, but you have pyparsing 3.0.8 which is incompatible.`

any recommendations?

**Comment by mkb79** [comment]
Hi @giovannicoppola me is fine, I hope you too! 

The only dependencies in my project which requires `pyparsing` is the [packaging](https://github.com/pypa/packaging) package. This need `pyparsing>=2.0.2,!=3.0.5`! So you can add `pyparsing<3,>=2.4.2` to your `requirement.txt`! This should solve your issue!

**Reply by mkb79** [reply]
Your `requirements.txt` should look like

```
pyparsing<3,>=2.4.2
audible-cli
```

**Reply by giovannicoppola** [reply]
thanks!! that installed `audible-cli` locally. Now, how to launch? For other packages I just added 
`export PYTHONPATH="$PWD/lib"` before launching my script, but that doesn't seem to work here... note that I launch `audible-cli` from within my python script like this (which might be overcomplicated but it works)
```
import os
os.system ("audible library export --format 'json'") 
```

**Reply by mkb79** [reply]
The `audible` entrypoint script is located in `lib/bin`. You can try `export PATH="$PWD/lib/bin"`! Please take care that I wrote `PATH` instead of `PYTHONPATH` which should point to `$PWD/lib`!

But you can write a shell script to run the `audible` command instead of a Python script.

**Reply by giovannicoppola** [reply]
this works!! I added `lib/bin` to my existing `export PATH` statement:
`export PATH=/opt/homebrew/bin:/usr/local/bin:$PATH:$PWD/lib/bin`

I launch it in Python because everything else downstream is in a Python script so it is easier for me. 
Thanks!! One last question (for now 😃): I am also adding the api catalog search and I was wondering: is it possible to add to your wishlist a new title you found with the catalog search? If so, I can implement as one of the options after catalog search.

thanks again for your patience.

**Reply by mkb79** [reply]
You can add an item to wishlist with `audible api -m post  -b '{"asin":"B004UZS0S0"}' wishlist`. `B004UZS0S0` is the asin of your book!

**Reply by mkb79** [reply]
Take a look [here](https://audible.readthedocs.io/en/latest/misc/external_api.html#post--1.0-wishlist)

**Reply by mkb79** [reply]
Or [here](https://audible.readthedocs.io/en/latest/misc/external_api.html#delete--1.0-wishlist-(string-asin)) for remove an item from the wishlist!

**Reply by giovannicoppola** [reply]
thanks @mkb79, I got adding to wishlist to work, but I don't seem to be able to find the corresponding remove command (e.g. corresponding to `audible api -m post -b '{"asin":"B004UZS0S0"}' wishlist` to remove `B004UZS0S0`, could you point me to it?

**Reply by mkb79** [reply]
The corresponding command to remove the item is `audible api -m delete wishlist/B004UZS0S0`. 

But there is an issue with audible v0.8.0 and the `delete` method of the `Client` class. This is solved in v0.8.1 which is published yesterday on PyPI. But `audible-cli` is pinning v0.8.0.

**Reply by mkb79** [reply]
Until I update `audible-cli` you can use the development branch in the meantime. Simply replace `audible-cli` in your `requirements.txt` with `audible-cli @ git+https://github.com/mkb79/audible-cli@development` as a workaround.

**Reply by mkb79** [reply]
The development branch has the `add` and `remove` subcommand in the `wishlist` build-in. So you don’t need to use the `api` command for adding or removing an item on the wishlist.

**Reply by giovannicoppola** [reply]
thank you, that works. Almost there! One more thing to check: when querying the catalog using `audible api`, I can't find the `cover_url` attribute, which is returned when I list `library` and `wishlist` with `audible`, is that accurate? This is the use case: if a user adds from the catalog to the wishlist, I wanted to fetch the cover image without rebuilding the whole database. 

BTW, I posted the draft of the [readme doc](https://github.com/giovannicoppola/alfred-audible), in case you have feedback. I will replace the GIF. If everything works I will make it public tomorrow!

**Reply by mkb79** [reply]
You have to add `media` to your `response_groups`. Then each product have an `product_images` key which has the cover url. 

FYI:
There are differences between the endpoints. Requesting the library endpoint will give you items with all keys. But the value is `None` if you don’t provide the right response group. The catalog endpoint on the other side will only return the keys for the items which are in the provided response groups. 

> BTW, I posted the draft of the [readme doc](https://github.com/giovannicoppola/alfred-audible), in case you have feedback. I will replace the GIF. If everything works I will make it public tomorrow!

Will look later this day and will report!

**Reply by giovannicoppola** [reply]
FYI I saw this warning, not sure if it comes from my code or the library:
`lib/audible_cli/decorators.py:181: SyntaxWarning: "is" with a literal. Did you mean "=="?
  if value is 0:`

**Reply by giovannicoppola** [reply]
I am testing on a different computer and it seems I did not bundle everything? Or it's because it is an M1 mac (I get a lot of `mach-o, but wrong architecture` error messages when I try to run `audible-quickstart`. I might create 2 builds, one for older Macs and one for M1s.

**Reply by mkb79** [reply]
> FYI I saw this warning, not sure if it comes from my code or the library

Thank you for your information. That’s my code which gives you the SyntaxWarning. 

> I am testing on a different computer and it seems I did not bundle everything? Or it's because it is an M1 mac (I get a lot of mach-o, but wrong architecture error messages when I try to run audible-quickstart. I might create 2 builds, one for older Macs and one for M1s.

Sorry I can’t help you by this issue. I don’t have a M1!

**Reply by giovannicoppola** [reply]
Unfortunately while everything works on my computer, but I can't get to install on a fresh computer, now I get permission issues
<img width="225" alt="image" src="https://user-images.githubusercontent.com/5774411/164815504-ca45bbec-3eaa-417c-a07c-507d9a341770.png">

**Reply by giovannicoppola** [reply]
`/lib/PIL/_imaging.cpython-38-darwin.so' not valid for use in process: library load disallowed by system policy)`

**Reply by giovannicoppola** [reply]
OK, I think I can use this solution for now: require the user to install `pip3` and install everything else via `pip3` the first time the app runs. this seems to work.

**Reply by mkb79** [reply]
> OK, I think I can use this solution for now: require the user to install pip3 and install everything else via pip3 the first time the app runs. this seems to work.

Yeah, I think it’s the best solution when the user install the packages itself. Or you use Pyinstaller to build a standalone binary and deliver this with your project.
